entity-registry/src/main/java/com/linkedin/metadata/aspect/validation/ConditionalWriteValidator.java

```diff
@@ -0,0 +1,225 @@
+ * prevent performing the write operation.
+ *
+ * <p>2. Headers are included with the MCP for the following time-based checks * If-Modified-Since *
+ * If-Unmodified-Since
+ */
+@Setter
+@Getter
+@Accessors(chain = true)
+public class ConditionalWriteValidator extends AspectPayloadValidator {
+  public static final String DEFAULT_ASPECT_VERSION = "1";
+  public static final long DEFAULT_LAST_MODIFIED_TIME = Long.MIN_VALUE;
+  public static final String HTTP_HEADER_IF_VERSION_MATCH = "If-Version-Match";
+
+  @Nonnull private AspectPluginConfig config;
+
+  private static boolean hasTimePrecondition(ChangeMCP item) {
+    return item.getHeader(HttpHeaders.IF_MODIFIED_SINCE).isPresent()
+        || item.getHeader(HttpHeaders.IF_UNMODIFIED_SINCE).isPresent();
+  }
+
+  private static boolean hasVersionPrecondition(ChangeMCP item) {
+    return item.getHeader(HTTP_HEADER_IF_VERSION_MATCH).isPresent();
+  }
+
+  private static boolean isApplicableFilter(ChangeMCP item) {
+    if (ChangeType.RESTATE.equals(item.getChangeType())) {
+      return false;
+    }
+
+    return hasTimePrecondition(item) || hasVersionPrecondition(item);
+  }
+
+  @Override
+  protected Stream<AspectValidationException> validatePreCommitAspects(
+      @Nonnull Collection<ChangeMCP> changeMCPs, @Nonnull RetrieverContext retrieverContext) {
+    ValidationExceptionCollection exceptions = ValidationExceptionCollection.newCollection();
+    AspectRetriever aspectRetriever = retrieverContext.getAspectRetriever();
+
+    List<ChangeMCP> applicableMCPs =
+        changeMCPs.stream()
+            .filter(ConditionalWriteValidator::isApplicableFilter)
+            .collect(Collectors.toList());
+
+    // Batch lookup missing data
+    Map<Urn, Set<String>> missingDataUrnAspects =
+        applicableMCPs.stream()
+            // create change types not expected to have previous data
+            .filter(
+                item ->
+                    !Set.of(ChangeType.CREATE, ChangeType.CREATE_ENTITY)
+                        .contains(item.getChangeType()))
+            .filter(item -> item.getPreviousSystemAspect() == null)
+            .collect(
+                Collectors.groupingBy(
+                    ChangeMCP::getUrn,
+                    Collectors.mapping(ChangeMCP::getAspectName, Collectors.toSet())));
+    final Map<Urn, Map<String, SystemAspect>> resolvedData =
+        aspectRetriever.getLatestSystemAspects(missingDataUrnAspects);
+
+    for (ChangeMCP item : applicableMCPs) {
+      // Validate aspect version precondition
+      if (hasVersionPrecondition(item)) {
+        item.getHeader(HTTP_HEADER_IF_VERSION_MATCH)
+            .flatMap(
+                headerValue ->
+                    validateVersionPrecondition(
+                        item, Pair.of(HTTP_HEADER_IF_VERSION_MATCH, headerValue), resolvedData))
+            .ifPresent(exceptions::addException);
+      }
+
+      // Validate modified time conditional
+      if (hasTimePrecondition(item)) {
```

Comment:
It is possible for a client to pass both a timestamp and version header, and this code will apply both checks.
Should we just prefer one over the other rather than applying both?

Reply:
From the way I read the http specification, both headers are allowed at the same time. It actually allows you to specify a time range when using both.

Reply:
While not implemented here, these headers can be used to time travel on a GET requests for example. Give me objects that were modified before/after or within the given time range specified by the headers.

Reply:
That's actually an interesting use-case

@david-leifker if i specify some time range, will it give me all versions that were created in that time range? or just the latest in that time range?

Reply:
Per the specification, it would return versions in that range.

Reply:
interesting! @ajoymajumdar this would be able to replace our queries against audit